test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -549,41 +546,61 @@ static URI startZookeeperInstance() {
         return zkUris.get(0);
     }
 
-    static void startBookkeeperInstances(final URI zkUri) {
-        Service bkService = new BookkeeperService("bookkeeper", zkUri);
+    static void startBookkeeperInstances(final URI zkUri) throws ExecutionException {
+        Service bkService = Utils.createBookkeeperService(zkUri);
         if (!bkService.isRunning()) {
             bkService.start(true);
         }
         List<URI> bkUris = bkService.getServiceDetails();
         log.debug("Bookkeeper service details: {}", bkUris);
     }
 
-    static URI startPravegaControllerInstances(final URI zkUri) {
-        Service controllerService = new PravegaControllerService("controller", zkUri);
+    static URI startPravegaControllerInstances(final URI zkUri) throws ExecutionException {
+        Service controllerService = Utils.createPravegaControllerService(zkUri);
         if (!controllerService.isRunning()) {
             controllerService.start(true);
         }
-        controllerService.scaleService(3, true);
+        Futures.getAndHandleExceptions(controllerService.scaleService(3), ExecutionException::new);
         List<URI> conUris = controllerService.getServiceDetails();
+        log.info("conuris {} {}", conUris.get(0), conUris.get(1));
         log.debug("Pravega Controller service  details: {}", conUris);
         // Fetch all the RPC endpoints and construct the client URIs.
-        final List<String> uris = conUris.stream().filter(uri -> uri.getPort() == 9092).map(URI::getAuthority)
-                .collect(Collectors.toList());
+        List<String> uris;
+        if (Utils.isDockerLocalExecEnabled()) {
```

Comment:
It is a bit odd that the test case needs to know whether it is running on docker or something else. Any specific behavior so be abstracted away from the test cases.

Reply:
So here, the test case, we try to access controller using both `tcp://` and `pravega://`  networks. And in the docker based execution ,  the communication works only in `tcp` network.
so , docker based execution tries to access controller only on `tcp://`

Reply:
I understand, but it doesn't make sense to ask developers of test cases to implement environment specific functionality. This should be abstracted away in some helper class .

Reply:
Created a parent and two sub classes for the same.